specreduce/tracing.py

```diff
@@ -110,3 +114,177 @@ def __post_init__(self):
         self._bound_trace()
+
+
+@dataclass
+class KosmosTrace(Trace):
+    """
+    Trace the spectrum aperture in an image.
+
+    Chops image up in bins along the dispersion (wavelength) direction,
+    fits a Gaussian within each bin to determine the trace's spatial
+    center. Finally, draws a cubic spline through the bins to up-sample
+    trace along every pixel in the dispersion direction.
+
+    (The original version of this algorithm is sourced from James
+     Davenport's `kosmos` repository.)
+
+    Parameters
+    ----------
+    image : `~astropy.nddata.NDData` or array-like, required
+        The image over which to run the trace. Assumes cross-dispersion
+        (spatial) direction is axis 0 and dispersion (wavelength)
+        direction is axis 1.
+    bins : int, optional
+        The number of bins in the dispersion (wavelength) direction
+        into which to divide the image. Use fewer if KosmosTrace is
+        having difficulty, such as with faint targets.
+        Minimum bin size is 4. [default: 20]
+    guess : int, optional
+        A guess at the trace's location in the cross-dispersion
+        (spatial) direction. If set, overrides the normal max peak
+        finder. Good for tracing a fainter source if multiple traces
+        are present. [default: None]
+    window : int, optional
+        Fit the trace to a region with size `window * 2` around the
+        guess position. Useful for tracing faint sources if multiple
+        traces are present, but potentially bad if the trace is
+        substantially bent or warped. [default: None]
+
+    Improvements Needed
+    -------------------
+    1) switch to astropy models for Gaussian (done)
+    2) return info about trace width (?)
+    3) add re-fit trace functionality (or break off into other method)
+    4) add other interpolation modes besides spline, maybe via
+        specutils.manipulation methods?
+    """
+    bins: int = 20
+    guess: float = None
+    window: int = None
+    _cross_disp_axis = 0
+    _disp_axis = 1
+
+    def __post_init__(self):
+        # handle multiple image types and mask uncaught invalid values
+        if isinstance(self.image, NDData):
+            img = np.ma.masked_invalid(np.ma.masked_array(self.image.data,
+                                                          mask=self.image.mask))
+        else:
+            img = np.ma.masked_invalid(self.image)
+
+        # validate arguments
+        if img.mask.all():
+            raise ValueError('image is fully masked. Check for invalid values')
+
+        if self._cross_disp_axis != 0:
+            raise ValueError('cross-dispersion axis must equal 0')
+
+        if self._disp_axis != 1:
+            raise ValueError('dispersion axis must equal 1')
+
+        if not isinstance(self.bins, int):
+            warnings.warn('TRACE: Converting bins to int')
+            self.bins = int(self.bins)
+
+        if self.bins < 4:
+            raise ValueError('bins must be >= 4')
```

Comment:
the useful upper limit is the length of `_disp_axis`. if i'm correctly reading the logic of where `x_bins` is being set, it might actually be `_disp_axis - 1`. if you don't want to add the check now, we can make an issue and tackle it later.

Reply:
I added this to the latest commit.